### ISSUE TITLE
[jvm-packages] evalSet returning constant value for all iterations.

### ISSUE THREAD

**Issue by mhmoudr** [dmlc/xgboost#2766]
For bugs or installation issues, please provide the following information.
The more information you provide, the more easily we will be able to offer
help and advice.

## Environment info
Operating System: Linux 
Package used (python/R/jvm/C++): jvm
`xgboost` version used: latest
The commit hash (`git rev-parse HEAD`): 178517524fd983ff1672d8929bd68ecfb4a0d730

## Steps to reproduce

1. try to build a tree model on a training dataset 
2. try to call evalSet for all iterations on hold-out dataset using code similar to:

```
(0 to iterations).foreach(i => {
    val evals = model.evalSet(Array(holdoutData), Array("holdout"), iter = i)
    println(evals)
})
```
then the result will be something similar to: 
```
[0]	holdout-rmse:22.571465
[1]	holdout-rmse:22.571465
[2]	holdout-rmse:22.571465
[3]	holdout-rmse:22.571465
[4]	holdout-rmse:22.571465
[5]	holdout-rmse:22.571465
[6]	holdout-rmse:22.571465
[7]	holdout-rmse:22.571465
[8]	holdout-rmse:22.571465
```
while I am expecting each iteration to give a different error, as I have tried to calculate the error manually and error was varying, also worth to mention that the error provided above was matching the first iteration error. 

## Expected result 
using the following code to calculate error for each iteration "not efficient enough as this code re calculate predictions from scratch"

```
(0 to iterations).foreach(i => {
    val lbl = holdoutData.getLabel
    val preds = model.predict(holdoutData,treeLimit = i).map(_(0))
    val mse = lbl.zip(preds).map(v=>Math.pow(v._1-v._2,2)).sum / lbl.length.toDouble
    val rmse = Math.sqrt(mse)
    println(s"$i: $rmse")
    rmse
})
```

Result:

```
0 :22.571464303032133
1 :22.84230245222776
2 :22.840524887185357
3 :22.838731954973362
4 :22.836923434431508
5 :22.83509939134127
6 :22.833259625287162
7 :22.831403881749665
8 :22.829532237854362
9 :22.827644608033697
10 :22.825740744783133
```

**Comment by mhmoudr** [comment]
probably this issue is back to the code that exists in learner.cc as EvalOneIter does not do any prediction relates to the provided iteration, my reading that iter parameter is used for logging only (correct me if I am wrong!)
```
std::string EvalOneIter(int iter, const std::vector<DMatrix *> &data_sets,
                          const std::vector<std::string> &data_names) override {
    double tstart = dmlc::GetTime();
    std::ostringstream os;
    os << '[' << iter << ']' << std::setiosflags(std::ios::fixed);
    if (metrics_.size() == 0) {
      metrics_.emplace_back(Metric::Create(obj_->DefaultEvalMetric()));
    }
    for (size_t i = 0; i < data_sets.size(); ++i) {
      this->PredictRaw(data_sets[i], &preds_);
      obj_->EvalTransform(&preds_);
      for (auto &ev : metrics_) {
        os << '\t' << data_names[i] << '-' << ev->Name() << ':'
           << ev->Eval(preds_, data_sets[i]->info(), tparam.dsplit == 2);
      }
    }

    if (tparam.debug_verbose > 0) {
      LOG(INFO) << "EvalOneIter(): " << dmlc::GetTime() - tstart << " sec";
    }
    return os.str();
  }
```

**Comment by mhmoudr** [comment]
I would say there is a defect in the code, where the following line 
```
this->PredictRaw(data_sets[i], &preds_);
```
should be 
```
this->PredictRaw(data_sets[i], &preds_, static_cast<unsigned int>(iter));
```
@CodingCat can you confirm ?

**Comment by khotilov** [comment]
EvalOneIter is simply a shortcut wrapper to be used *during* training, and it is not supposed to be used after training to obtain iteration-by-iteration evaluations. And you are correct that the iter parameter is used for logging only.

Other facts to consider:
- in general, ntree_limit is not the same as iteration number, so you cannot do what you are proposing here and in #3108
- even if you would properly define the ntree_limit in there
  - such modification would still break things, such as caching as wellas the DART booster
  - there would be no advantage WRT calling `predict`, since both are calling PredictRaw, and the prediction cache stores only the last predictions, not the whole history

So you have several options:
- the best option would be to collect the evaluations during training
- but if it's "too late", using predict with ntree_limit is ok
- also you may predict the tree leaves, map them to leaf values, and manually compute cumulative iteration-by-iteration predictions.

**Comment by mhmoudr** [comment]
Thenks @khotilov for clarifications, now I can see the issue...

While on the other hand I would argue the need of exposing such function `evalSet` to the model API, I think it is confusing to have it there. 

And back to the options you have suggested, it is always beneficial to be able to evaluate model trees after training, on different dataset, which eliminate the first option. 

And in term of predicting with `ntree_limit` logically it solves the problem but performance wise it was too bad, as it takes several hours to do such operation for 2000 tree model, as calculation for early trees will be repeated every time you call the next one, assuming complicity to be O(N^2).

Third option sounds the best, as predicting each tree will happen once and values can be aggregated cumulatively, but could not find any API on the model that expose predicting tree leaves (in JVM packages) so probably we need to either add a c function to do the whole calculation in native code, or we can just expose model trees to be able to call `GetLeafIndex` and `leaf_value`  on each tree, thoughts!?
